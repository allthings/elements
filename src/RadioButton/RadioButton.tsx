import { css } from 'glamor'
import React, { ChangeEvent, Component } from 'react'
import Absolute from '../Absolute'
import Circle from '../Circle'
import Relative from '../Relative'
import Text from '../Text'
import View from '../View'

const styles = {
  radioElement: css({
    marginRight: 40,
    width: 'auto',
  }),
  radio: css({
    opacity: 0,
    margin: 0,
  }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
}

export interface IRadioButtonProps {
  readonly id?: string
  /** Set to true to controll radio button */
  readonly checked?: boolean
  /** The name of this input field */
  readonly name?: string
  /** Called when a radio button is clicked */
  readonly onChange?: (e: ChangeEvent) => void
  /** Mark if the RadioButton is required */
  readonly required?: boolean
  /** The value the checkbox will have */
  readonly value: string
}

interface IState {
  checked: boolean
}

/**
 * RadioButtonSet can be used to render a set of RadioButtons to allow users to select exactly one item from a set.
 * Like gender (male / female) or sizes (s,m,l,xl)
 *
 * ```example
 * <RadioButtonSet name="gender" defaultValue="male" required>
 *   <RadioButton value="female">Female</RadioButton>
 *   <RadioButton value="male">Male</RadioButton>
 * </RadioButtonSet>
 * ```
 */
class RadioButton extends Component<IRadioButtonProps, IState> {
  static defaultProps = {
    required: false,
  }

  state = {
    // set initial value to boolean, to avoid the warning:
    // `A component is changing an uncontrolled input of type radio to be controlled.`
    checked: !!this.props.checked,
  }

  static getDerivedStateFromProps(props: IRadioButtonProps) {
    if (props.checked === true) {
      return {
        checked: true,
      }
    } else if (props.checked === false) {
      return {
        checked: false,
      }
    } else {
      return null
    }
  }

  handleChange = (e: ChangeEvent) => {
    this.props.onChange && this.props.onChange(e)
    this.setState(({ checked }) => ({ checked: !checked }))
  }

  render() {
    const {
      children,
      id,
      name,
      value,
      checked,
      onChange,
      ...props
    } = this.props

    const realId = id || value

    return (
      <View
        {...styles.radioElement}
        {...(props as Omit<IRadioButtonProps, 'checked' | 'onChange'>)}
      >
        <Relative direction="row" alignV="center" alignH="center">
          <View htmlElement="label">
            <Relative top={10}>
              <Circle
                color={this.state.checked ? 'primary' : 'lightGrey'}
                radius={20}
              >
                <Circle color="white" radius={16}>
                  <Circle
                    color={this.state.checked ? 'primary' : 'white'}
                    radius={10}
                  />
                </Circle>
              </Circle>
            </Relative>
            <Relative
              top={-8}
              style={{
                marginLeft: 30,
              }}
            >
              {typeof children === 'string' ? (
                <Text size="m">{children}</Text>
              ) : (
                children
              )}
            </Relative>
            <Absolute top={0} left={0} right={0} bottom={0}>
              <input
                id={realId}
                type="radio"
                name={name}
                {...styles.radio}
                value={value}
                checked={this.state.checked}
                onChange={this.handleChange}
              />
            </Absolute>
          </View>
        </Relative>
      </View>
    )
  }
}

export default RadioButton
