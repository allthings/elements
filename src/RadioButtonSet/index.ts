export { default } from './RadioButtonSet'
export * from './RadioButtonSet'
