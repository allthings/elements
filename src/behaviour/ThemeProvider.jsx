import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
