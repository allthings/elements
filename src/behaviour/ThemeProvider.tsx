import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
