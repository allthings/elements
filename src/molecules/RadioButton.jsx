import React from 'react'
import PropTypes from 'prop-types'
import { css } from 'glamor'
import Absolute from '../atoms/Absolute'
import Circle from '../atoms/Circle'
import Relative from '../atoms/Relative'
import Text from '../atoms/Text'
import View from '../atoms/View'
import Theme from '../behaviour/Theme'

const styles = {
  radioElement: css({
    marginRight: 40,
    width: 'auto',
  }),
  radio: primaryColor =>
    css({
      opacity: 0,
      margin: 0,
    }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
}

/**
 * RadioButtonSet can be used to render a set of RadioButtons to allow users to select exactly one item from a set.
 * Like gender (male / female) or sizes (s,m,l,xl)
 *
 * ```example
 *
 * const gender = [
 *    { key: m, value: 'male' },
 *    { key: f, value: 'female' },
 * ]
 *
 * <RadioButtonSet name="gender" defaultValue="male" required />
 *
 * <RadioButton
 *    name="sizes"
 *    defaultValue="large"
 *    required
 * />
 * ```
 */
class RadioButton extends React.Component {
  static propTypes = {
    /** The value the checkbox will have */
    value: PropTypes.string.isRequired,
    /** The name of this input field */
    name: PropTypes.string,
    /** Called when a radio button is clicked */
    onChange: PropTypes.func,
    /** Background color of the form item */
    backgroundColor: PropTypes.string,
    /** Background color of the form item */
    id: PropTypes.string,
    /** Mark if the RadioButton is required */
    required: PropTypes.bool,
    /** Set to true to controll radio button */
    checked: PropTypes.bool,
    /** Label of the radio button */
    children: PropTypes.node,
    /** The referenced input field */
    inputRef: PropTypes.inputRef,
  }

  static defaultProps = {
    required: false,
  }

  state = {
    checked: undefined,
  }

  static getDerivedStateFromProps(props) {
    if (props.checked === true) {
      return {
        checked: true,
      }
    } else if (props.checked === false) {
      return {
        checked: false,
      }
    } else {
      return null
    }
  }

  handleChange = e => {
    console.log('ok', this.props)
    this.props.onChange && this.props.onChange(e)
    this.setState(({ checked }) => ({ checked: !checked }))
  }

  render() {
    const {
      backgroundColor,
      inputRef,
      onChange,
      name,
      id,
      children,
      checked,
      value,
      ...props
    } = this.props

    const realId = id || value

    return (
      <Theme>
        {({ theme, colorize }) => (
          <View {...styles.radioElement} {...props}>
            <Relative direction="row" alignV="center" alignH="center">
              <View htmlElement="label" htmlFor={realId}>
                <Relative top={10}>
                  <Circle
                    color={`${
                      this.state.checked ? theme.primary : 'lightGrey'
                    }`}
                    radius={20}
                  >
                    <Circle color="white" radius={16}>
                      <Circle
                        color={`${
                          this.state.checked ? theme.primary : 'white'
                        }`}
                        radius={10}
                      />
                    </Circle>
                  </Circle>
                </Relative>
                <Relative
                  top={-8}
                  style={{
                    marginLeft: 30,
                  }}
                >
                  {typeof children === 'string' ? (
                    <Text size="m">{children}</Text>
                  ) : (
                    children
                  )}
                </Relative>
                <Absolute top={0} left={0} right={0} bottom={0}>
                  <input
                    id={realId}
                    type="radio"
                    ref={inputRef}
                    name={name}
                    {...styles.radio(theme.primary)}
                    value={value}
                    checked={this.state.checked}
                    onChange={this.handleChange}
                  />
                </Absolute>
              </View>
            </Relative>
          </View>
        )}
      </Theme>
    )
  }
}

export default RadioButton
