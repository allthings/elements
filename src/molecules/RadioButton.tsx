import React, { Component, ChangeEvent, PropsWithChildren } from 'react'
import { css } from 'glamor'
import Absolute from '../atoms/Absolute'
import Circle from '../atoms/Circle'
import Relative from '../atoms/Relative'
import Text from '../atoms/Text'
import View from '../atoms/View'
import Theme from '../behaviour/Theme'

const styles = {
  radioElement: css({
    marginRight: 40,
    width: 'auto',
  }),
  radio: css({
    opacity: 0,
    margin: 0,
  }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
}

export interface IRadioButtonProps {
  /** Background color of the form item */
  backgroundColor?: string
  id?: string
  /** Set to true to controll radio button */
  checked?: boolean
  /** The name of this input field */
  name?: string
  /** Called when a radio button is clicked */
  onChange?: (e: ChangeEvent) => void
  /** Mark if the RadioButton is required */
  required?: boolean
  /** The value the checkbox will have */
  value: string
}

interface IState {
  checked: boolean
}

/**
 * RadioButtonSet can be used to render a set of RadioButtons to allow users to select exactly one item from a set.
 * Like gender (male / female) or sizes (s,m,l,xl)
 *
 * ```example
 * <RadioButtonSet name="gender" defaultValue="male" required>
 *   <RadioButton value="female">Female</RadioButton>
 *   <RadioButton value="male">Male</RadioButton>
 * </RadioButtonSet>
 * ```
 */
class RadioButton extends Component<
  PropsWithChildren<IRadioButtonProps>,
  IState
> {
  static defaultProps = {
    required: false,
  }

  state = {
    // set initial value to boolean, to avoid the warning:
    // `A component is changing an uncontrolled input of type radio to be controlled.`
    checked: !!this.props.checked,
  }

  static getDerivedStateFromProps(props: IRadioButtonProps) {
    if (props.checked === true) {
      return {
        checked: true,
      }
    } else if (props.checked === false) {
      return {
        checked: false,
      }
    } else {
      return null
    }
  }

  handleChange = (e: ChangeEvent) => {
    this.props.onChange && this.props.onChange(e)
    this.setState(({ checked }) => ({ checked: !checked }))
  }

  render() {
    const {
      backgroundColor,
      checked,
      children,
      id,
      name,
      onChange,
      value,
      ...props
    } = this.props

    const realId = id || value

    return (
      <Theme>
        {({ theme }) => (
          <View {...styles.radioElement} {...props}>
            <Relative direction="row" alignV="center" alignH="center">
              <View htmlElement="label">
                <Relative top={10}>
                  <Circle
                    color={`${
                      this.state.checked ? theme.primary : 'lightGrey'
                    }`}
                    radius={20}
                  >
                    <Circle color="white" radius={16}>
                      <Circle
                        color={`${
                          this.state.checked ? theme.primary : 'white'
                        }`}
                        radius={10}
                      />
                    </Circle>
                  </Circle>
                </Relative>
                <Relative
                  top={-8}
                  style={{
                    marginLeft: 30,
                  }}
                >
                  {typeof children === 'string' ? (
                    <Text size="m">{children}</Text>
                  ) : (
                    children
                  )}
                </Relative>
                <Absolute top={0} left={0} right={0} bottom={0}>
                  <input
                    id={realId}
                    type="radio"
                    name={name}
                    {...styles.radio}
                    value={value}
                    checked={this.state.checked}
                    onChange={this.handleChange}
                  />
                </Absolute>
              </View>
            </Relative>
          </View>
        )}
      </Theme>
    )
  }
}

export default RadioButton
