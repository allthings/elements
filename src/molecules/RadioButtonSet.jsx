import React from 'react'
import PropTypes from 'prop-types'
import { css } from 'glamor'
import Absolute from '../atoms/Absolute'
import Circle from '../atoms/Circle'
import Relative from '../atoms/Relative'
import Text from '../atoms/Text'
import View from '../atoms/View'
import ListItem from './List/ListItem'
import Theme from '../behaviour/Theme'

const styles = {
  radioWrapper: () =>
    css({
      padding: '15px 15px',
    }),
  radioElement: num =>
    css({
      marginRight: 40,
      width: 'auto',
    }),
  radio: primaryColor =>
    css({
      opacity: 0,
      margin: 0,
    }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
}

/**
 * RadioButtonSet can be used to render a set of RadioButtons to allow users to select exactly one item from a set.
 * Like gender (male / female) or sizes (s,m,l,xl)
 *
 * ```example
 *
 * const gender = [
 *    { key: m, value: 'male' },
 *    { key: f, value: 'female' },
 * ]
 *
 * <RadioButtonSet name="gender" selection={gender} defaultValue="male" required />
 *
 *
 * const sizes = [
 *    { key: s, value: 'short' },
 *    { key: m, value: 'medium' },
 *    { key: l, value: 'large' },
 *    { key: xl, value: 'x-large' },
 * ]
 *
 * <RadioButtonSet
 *    name="sizes"
 *    selection={sizes}
 *    defaultValue="large"
 *    radioSetStyles={css({
 *                   margin: '0 50px',
 *                 })}
 *    radioElementStyles={css({
 *                   width: '100%',
 *                   backgroundColor: '#f00',
 *                 })}
 *    required
 * />
 * ```
 */
class RadioButtonSet extends React.Component {
  static propTypes = {
    /** The default value to put into the component, without making it controlled */
    defaultValue: PropTypes.string,
    /** The name of this input field */
    name: PropTypes.string.isRequired,
    /** Called when a radio button is clicked */
    onChange: PropTypes.func,
    /** an array of key/values */
    selection: PropTypes.array.isRequired,
    /** Background color of the form item */
    backgroundColor: PropTypes.string,
    /** Text size of the label one of xs,s,m,l,xl */
    labelSize: Text.propTypes.size,
    /** css object, that overwrites the default styles of the radio set */
    radioSetStyles: PropTypes.object,
    /** css object, that overwrites the default styles of a single radio element */
    radioElementStyles: PropTypes.object,
  }

  static defaultProps = {
    required: false,
    labelSize: 'l',
  }

  state = {
    visible: true,
    selected: this.props.defaultValue,
  }

  handleChange = e => {
    this.props.onChange && this.props.onChange(e)
    this.setState({ selected: e.target.value })
  }

  isChecked = checked => {
    return checked ? { defaultChecked: 'checked' } : {}
  }

  render() {
    const {
      backgroundColor,
      selection,
      defaultValue,
      onChange,
      labelSize,
      radioSetStyles,
      radioElementStyles,
      ...props
    } = this.props
    return (
      <Theme>
        {({ theme, colorize }) => (
          <ListItem
            alignV="start"
            padded={false}
            backgroundColor={colorize(backgroundColor)}
            direction={`${selection.length > 2 ? 'column' : 'row'}`}
            {...styles.radioWrapper(selection.length)}
            {...radioSetStyles}
          >
            {selection.map(sel => {
              const realChecked = sel.value === this.state.selected
              return (
                <View
                  key={sel.key}
                  {...styles.radioElement(selection.length)}
                  {...radioElementStyles}
                >
                  {
                    <Relative direction="row" alignV="center" alignH="center">
                      <label htmlFor={sel.value}>
                        <Relative top={10}>
                          <Circle
                            color={`${
                              realChecked ? theme.primary : 'lightGrey'
                            }`}
                            radius={20}
                          >
                            <Circle color="white" radius={16}>
                              <Circle
                                color={`${
                                  realChecked ? theme.primary : 'white'
                                }`}
                                radius={10}
                              />
                            </Circle>
                          </Circle>
                        </Relative>
                        <Relative
                          top={-10}
                          style={{
                            marginLeft: 30,
                          }}
                        >
                          <Text size={labelSize}>{sel.value}</Text>
                        </Relative>
                        <Absolute top={0} left={0} right={0} bottom={0}>
                          <input
                            id={sel.value}
                            type="radio"
                            name="title"
                            {...styles.radio(theme.primary)}
                            value={sel.value}
                            {...props}
                            {...this.isChecked(realChecked)}
                            onChange={this.handleChange}
                          />
                        </Absolute>
                      </label>
                    </Relative>
                  }
                </View>
              )
            })}
          </ListItem>
        )}
      </Theme>
    )
  }
}

RadioButtonSet.contextTypes = {
  validity: PropTypes.object,
  STATES: PropTypes.array,
}

export default RadioButtonSet
