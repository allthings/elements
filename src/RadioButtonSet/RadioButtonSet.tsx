import { css } from 'glamor'
import React, {
  ChangeEvent,
  Children,
  cloneElement,
  Component,
  Fragment,
  isValidElement,
  ReactElement,
} from 'react'
import { IRadioButtonProps } from '../RadioButton'
import Text from '../Text'
import View from '../View'

const styles = {
  label: css({
    position: 'relative',
    marginBottom: 10,
  }),
}

const DIRECTION_HORIZONTAL = 'horizontal'
const DIRECTION_VERTICAL = 'vertical'
const DIRECTION_AUTO = 'auto'

type DirectionType = 'horizontal' | 'vertical' | 'auto'

interface IRadioButtonSetProps {
  /** The default value to put into the component, without making it controlled */
  defaultValue?: string
  /** Customize direction */
  direction?: DirectionType
  /** The label of this input field */
  label?: string
  /** The name of this input field */
  name: string
  /** Called when a radio button is clicked */
  onChange?: (e: ChangeEvent) => void
  /** Pass true to mark the field as required */
  required?: boolean
}

interface IState {
  value?: string
}

/**
 * RadioButtonSet can be used to render a set of RadioButtons to allow users to select exactly one item from a set.
 *
 * ```example
 * <RadioButtonSet
 *    name="order"
 *    defaultValue="tee"
 *    required
 * >
 *   <RadioButton value="coffe">Coffe</RadioButton>
 *   <RadioButton value="tee">Tee</RadioButton>
 *   <RadioButton value="others">Others</RadioButton>
 * </RadioButtonSet>
 * ```
 */
class RadioButtonSet extends Component<IRadioButtonSetProps, IState> {
  static DIRECTION_HORIZONTAL = DIRECTION_HORIZONTAL
  static DIRECTION_VERTICAL = DIRECTION_VERTICAL
  static DIRECTION_AUTO = DIRECTION_AUTO

  static defaultProps = {
    direction: DIRECTION_AUTO,
  }

  state = {
    value: this.props.defaultValue,
  }

  radios = []

  handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    this.props.onChange && this.props.onChange(e)
    this.setState({ value: e.target.value })
  }

  getAutoDirection = () =>
    Children.count(this.props.children) > 2 ? 'column' : 'row'

  getDirection = () =>
    this.props.direction === DIRECTION_HORIZONTAL ? 'row' : 'column'

  render() {
    const {
      children,
      label,
      required,
      name,
      direction,
      onChange,
      ...props
    } = this.props
    const { value } = this.state

    return (
      <View>
        {label && (
          <View
            className="label"
            {...styles.label}
            style={{
              opacity: 1,
              top: 8,
            }}
          >
            <Text color="secondaryText" size="xs">
              {label} {required && '*'}
            </Text>
          </View>
        )}
        <View
          alignH="space-between"
          alignV="start"
          style={{ width: '100%' }}
          direction={
            direction === DIRECTION_AUTO
              ? this.getAutoDirection()
              : this.getDirection()
          }
          {...props}
        >
          {Children.map(children, (child: ReactElement<IRadioButtonProps>) => {
            if (isValidElement(child)) {
              if (child.type === Fragment) {
                console.log(
                  [
                    "Elements: the RadioButtonSet component doesn't accept a Fragment as a child.",
                    'Consider providing an array instead.',
                  ].join('\n'),
                )
              }

              return cloneElement(child, {
                name,
                checked: value === child.props.value,
                onChange: this.handleChange,
              })
            }

            return null
          })}
        </View>
      </View>
    )
  }
}

export default RadioButtonSet
