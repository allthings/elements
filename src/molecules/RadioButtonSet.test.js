import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent } from 'react-testing-library'
import RadioButton from './RadioButton'
import RadioButtonSet from './RadioButtonSet'
import ThemeProvider from '../behaviour/ThemeProvider'

describe('<RadioButtonSet />', () => {
  it('should render without error', () => {
    const tree = renderer
      .create(
        <ThemeProvider>
          <RadioButtonSet name="colors">
            <RadioButton value="blue">Blue</RadioButton>
            <RadioButton value="gray">Gray</RadioButton>
          </RadioButtonSet>
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should chain the onChange property', () => {
    const spy = jest.fn()
    const { getByText } = render(
      <ThemeProvider>
        <RadioButtonSet name="colors" onChange={spy} default>
          <RadioButton value="blue">Blue</RadioButton>
          <RadioButton value="gray">Gray</RadioButton>
        </RadioButtonSet>
      </ThemeProvider>
    )

    const radio = getByText('Gray')
    radio.value = 'blue'
    fireEvent.change(radio)
    expect(spy).toHaveBeenCalledTimes(1)
  })
})
