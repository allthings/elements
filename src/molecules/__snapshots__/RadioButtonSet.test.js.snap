// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButtonSet /> should render without error 1`] = `
.css-0,
[data-css-0] {
  margin-right: 40px;
  width: auto;
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  align-content: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex: 0 0 auto;
  -webkit-align-content: flex-start;
  -webkit-box-align: start;
  -webkit-align-items: flex-start;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  opacity: 0;
  margin: 0;
}

.css-4,
[data-css-4] {
  position: relative;
}

.css-5,
[data-css-5] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

.css-6,
[data-css-6] {
  position: relative;
  top: 10px;
}

.css-7,
[data-css-7] {
  background-color: lightGrey;
  border-radius: 50%;
  height: 20px;
  width: 20px;
  transition: background-color 120ms linear;
  -webkit-transition: background-color 120ms linear;
  -moz-transition: background-color 120ms linear;
}

.css-8,
[data-css-8] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

.css-9,
[data-css-9] {
  background-color: white;
  border-radius: 50%;
  height: 16px;
  width: 16px;
  transition: background-color 120ms linear;
  -webkit-transition: background-color 120ms linear;
  -moz-transition: background-color 120ms linear;
}

.css-10,
[data-css-10] {
  background-color: white;
  border-radius: 50%;
  height: 10px;
  width: 10px;
  transition: background-color 120ms linear;
  -webkit-transition: background-color 120ms linear;
  -moz-transition: background-color 120ms linear;
}

.css-11,
[data-css-11] {
  position: relative;
  top: -8px;
}

.css-12,
[data-css-12] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 13px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-13,
[data-css-13] {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}

<div
  data-css-1=""
>
  <div
    data-css-2=""
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      data-css-0=""
      data-css-1=""
      required={false}
    >
      <div
        data-css-4=""
        data-css-5=""
      >
        <label
          data-css-1=""
        >
          <div
            data-css-6=""
            data-css-1=""
          >
            <div
              data-css-7=""
              data-css-8=""
            >
              <div
                data-css-9=""
                data-css-8=""
              >
                <div
                  data-css-10=""
                  data-css-8=""
                />
              </div>
            </div>
          </div>
          <div
            data-css-1=""
            data-css-11=""
            style={
              Object {
                "marginLeft": 30,
              }
            }
          >
            <div
              data-css-12=""
              data-css-1=""
            >
              Blue
            </div>
          </div>
          <div
            data-css-1=""
            data-css-13=""
          >
            <input
              checked={false}
              data-css-3=""
              id="blue"
              name="colors"
              onChange={[Function]}
              type="radio"
              value="blue"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      data-css-0=""
      data-css-1=""
      required={false}
    >
      <div
        data-css-4=""
        data-css-5=""
      >
        <label
          data-css-1=""
        >
          <div
            data-css-6=""
            data-css-1=""
          >
            <div
              data-css-7=""
              data-css-8=""
            >
              <div
                data-css-9=""
                data-css-8=""
              >
                <div
                  data-css-10=""
                  data-css-8=""
                />
              </div>
            </div>
          </div>
          <div
            data-css-1=""
            data-css-11=""
            style={
              Object {
                "marginLeft": 30,
              }
            }
          >
            <div
              data-css-12=""
              data-css-1=""
            >
              Gray
            </div>
          </div>
          <div
            data-css-1=""
            data-css-13=""
          >
            <input
              checked={false}
              data-css-3=""
              id="gray"
              name="colors"
              onChange={[Function]}
              type="radio"
              value="gray"
            />
          </div>
        </label>
      </div>
    </div>
  </div>
</div>
`;
